### ISSUE TITLE
Streamline the ITestObjectFactory usage

### ISSUE THREAD

**Issue by krmahadevan** [testng-team/testng#2679]
Currently TestNG's object factory instantiation and usage pattern in the codebase is a bit haywire. This needs to be streamlined from the implementation perspective so that, its clear as to when and under what circumstances would TestNG use a global ITestObjectFactory instance and when would it resort to one provided via a suite.

This issue is an outcome from the pull request https://github.com/cbeust/testng/pull/2677

**Comment by mr-possible** [comment]
Can this issue be closed now?

**Comment by krmahadevan** [comment]
@mr-possible No. i dont think we can close it. This issue is tracking some architecture discrepancies that are highlighted in the linked issue and I dont remember going back to fixing them. Need to recheck as to where we are on this one.

**Comment by juherr** [comment]
**Verdict: still relevant**

PR #2677 added object-factory injection via listeners, but it did not streamline the underlying model. Current `master` still has a global `TestNG.m_objectFactory`, may replace it from a suite-level `object-factory` during `initializeConfiguration()`, and `SuiteRunner` separately retains an `objectFactory` reference.

This is therefore still a valid architecture cleanup. The remaining work needs a maintainer decision on the precedence/lifetime contract before implementation.

No tests were run as part of this triage.
